### PR TITLE
feat: validation rule for type and relation name

### PR DESCRIPTION
**Pull request by adriantam**
## Description
Allow specifying regex rule for type and relation name validation

## Review Checklist
- [x] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: src/check-dsl.ts [hunk 1]
```diff
@@ -9,6 +10,11 @@ import {
 } from "./parse-dsl";
 import { report } from "./reporters";
 
+interface validationRegex {
```
Comment (rhamzeh) [inline src/check-dsl.ts:13]:
```suggestion
interface ValidationRegex {
```

File: src/check-dsl.ts [hunk 2]
```diff
@@ -47,6 +53,8 @@ function populateRelations(
   lines: string[],
   reporter: any,
   parserResults: ParserResult,
+  typeRegex: validationRegex,
```
Comment (rhamzeh) [inline src/check-dsl.ts:56]:
```suggestion
  typeRegex: ValidationRegex,
```

File: src/check-dsl.ts [hunk 3]
```diff
@@ -47,6 +53,8 @@ function populateRelations(
...
+  relationRegex: validationRegex,
```
Comment (rhamzeh) [inline src/check-dsl.ts:57]:
```suggestion
  relationRegex: ValidationRegex,
```

File: src/check-dsl.ts [hunk 4]
```diff
@@ -168,15 +184,34 @@ export const basicValidateRelation = (
   });
 };
 
-export const checkDSL = (codeInEditor: string) => {
+export const checkDSL = (
+  codeInEditor: string,
+  typeValidation: string = defaultTypeRule,
```
Comment (rhamzeh) [inline src/check-dsl.ts:189]: Can we pass these in options? As we add more things, the list will grow

Comment (rhamzeh) [reply src/check-dsl.ts:189]: We should validate that what is passed can be converted to a regex and throw an appropriate error if not

File: src/default-regex.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+// These rules should match the rules as specified in
+// https://github.com/openfga/api/blob/main/openfga/v1/openfga.proto
```
Comment (rhamzeh) [inline src/default-regex.ts:2]: Can you link to the appropriate part please? 🙏🏽

File: src/reporters.ts [hunk 6]
```diff
@@ -104,6 +104,17 @@ export const reportUseSelf = ({ markers, lines, lineIndex, value }: ReporterOpts
   });
 };
 
+export const reportInvalidName = ({ markers, lines, lineIndex, value, clause }: ReporterOpts) => {
```
Comment (rhamzeh) [inline src/reporters.ts:107]: This should probably be more specific. It should either say type XYZ or relation ABC of type XYZ does not match..

File: src/default-regex.ts (hunk 5 above)
Comment (adriantam) [reply src/default-regex.ts:2]: The problem is that we either have to link with respect to a specific version or the line numbers will become out of date

File: src/check-dsl.ts (hunk 4 above)
Comment (adriantam) [reply src/check-dsl.ts]: Done

File: src/reporters.ts (hunk 6 above)
Comment (adriantam) [reply src/reporters.ts]: Done

File: src/check-dsl.ts [hunk 7]
```diff
@@ -168,15 +189,56 @@ export const basicValidateRelation = (
   });
 };
 
-export const checkDSL = (codeInEditor: string) => {
+export const checkDSL = (codeInEditor: string, options?: ValidationOptions) => {
   const lines = codeInEditor.split("\n");
   const markers: any = [];
   const reporter = report({ lines, markers });
+  const typeValidation = options && options.typeValidation ? options.typeValidation : defaultTypeRule;
+  const relationValidation = options && options.relationValidation ? options.relationValidation : defaultRelationRule;
```
Comment (rhamzeh) [inline src/check-dsl.ts:197]:
```suggestion
  const relationValidation = options.relationValidation || defaultRelationRule;
```

**Review by rhamzeh — DISMISSED** [review]

**Review by rhamzeh — APPROVED** [review]

**Pull request merged**